corehq/apps/userreports/expressions/specs.py

```diff
@@ -304,3 +305,31 @@ def _get_forms(self, case_id, context):
 
         context.set_cache_value(cache_key, xforms)
         return xforms
+
+
+class SubcasesExpressionSpec(JsonObject):
+    type = TypeProperty('get_subcases')
+    case_id_expression = DefaultProperty(required=True)
+
+    def configure(self, case_id_expression):
+        self._case_id_expression = case_id_expression
+
+    def __call__(self, item, context=None):
+        case_id = self._case_id_expression(item, context)
+        if not case_id:
+            return []
+
+        assert context.root_doc['domain']
+        return self._get_subcases(case_id, context)
+
+    def _get_subcases(self, case_id, context):
+        domain = context.root_doc['domain']
+        cache_key = (self.__class__.__name__, case_id)
+        if context.get_cache_value(cache_key) is not None:
+            return context.get_cache_value(cache_key)
+
+        case_accessors = CaseAccessors(domain)
+        indices = case_accessors.get_all_reverse_indices_info([case_id])
+        subcases = [c.to_json() for c in case_accessors.get_cases([i.case_id for i in indices])]
```

Comment:
all good - it uncovered a few inconsistencies in the code!

On Tue, Jun 21, 2016 at 3:22 PM, Farid Rener <notifications@github.com>
wrote:

> In corehq/apps/userreports/expressions/specs.py
> <https://github.com/dimagi/commcare-hq/pull/12151#discussion_r67866281>:
>
> > +        case_id = self._case_id_expression(item, context)
> > +        if not case_id:
> > +            return []
> > +
> > +        assert context.root_doc['domain']
> > +        return self._get_subcases(case_id, context)
> > +
> > +    def _get_subcases(self, case_id, context):
> > +        domain = context.root_doc['domain']
> > +        cache_key = (self.__class__.__name__, case_id)
> > +        if context.get_cache_value(cache_key) is not None:
> > +            return context.get_cache_value(cache_key)
> > +
> > +        case_accessors = CaseAccessors(domain)
> > +        indices = case_accessors.get_all_reverse_indices_info([case_id])
> > +        subcases = [c.to_json() for c in case_accessors.get_cases([i.case_id for i in indices])]
>
> 😞 sorry about that.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/12151/files/4ec35605e6f69f69c6e55cf0246d040ca4a32211#r67866281>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAED-17NlRx7C8CYMV5vKg3XCLm5LDSuks5qN-WxgaJpZM4I57db>
> .
>